### PR TITLE
AWS S3 IO through KvikIO

### PR DESCRIPTION
**Pull request by madsbk**
Implement remote IO read using KvikIO's S3 backend. 

For now, this is an experimental feature for parquet read only. Enable by defining `CUDF_KVIKIO_REMOTE_IO=ON`.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/cmake/thirdparty/get_kvikio.cmake [hunk 1]
```diff
@@ -15,12 +15,14 @@
...
+  set(CPM_DOWNLOAD_KvikIO ON)
+
   rapids_cpm_find(
     KvikIO ${VERSION}
     GLOBAL_TARGETS kvikio::kvikio
     CPM_ARGS
-    GIT_REPOSITORY https://github.com/rapidsai/kvikio.git
-    GIT_TAG branch-${VERSION}
+    GIT_REPOSITORY https://github.com/madsbk/kvikio
+    GIT_TAG remote-io-s3
```
Comment (madsbk) [inline cpp/cmake/thirdparty/get_kvikio.cmake:25]: Remove when https://github.com/rapidsai/kvikio/pull/479 has been merged

File: cpp/cmake/thirdparty/get_kvikio.cmake [hunk 2]
```diff
@@ -15,12 +15,13 @@
...
 
+  set(CPM_DOWNLOAD_KvikIO ON)
   rapids_cpm_find(
     KvikIO ${VERSION}
     GLOBAL_TARGETS kvikio::kvikio
     CPM_ARGS
-    GIT_REPOSITORY https://github.com/rapidsai/kvikio.git
-    GIT_TAG branch-${VERSION}
+    GIT_REPOSITORY https://github.com/madsbk/kvikio.git
+    GIT_TAG remote-io-s3
```
Comment (madsbk) [inline cpp/cmake/thirdparty/get_kvikio.cmake:24]: Revert when https://github.com/rapidsai/kvikio/pull/479 has been merged

File: python/cudf/cudf/io/parquet.py [hunk 3]
```diff
@@ -622,6 +622,7 @@ def read_parquet(
         storage_options=storage_options,
         bytes_per_thread=bytes_per_thread,
         prefetch_options=prefetch_options,
+        kvikio_remote_io=cudf.get_option("kvikio_remote_io"),
```
Comment (rjzamora) [inline python/cudf/cudf/io/parquet.py:625]: My understanding is that this feature is independent of the file format (e.g. parquet). Can we avoid adding this argument to `ioutils.get_reader_filepath_or_buffer` for now and just rely on the `"kvikio_remote_io"` option/configuration within that function?

I suppose we *may* want a key-word argument eventually, but I don't see much reason to add it now (even your read_parquet test just sets the `"kvikio_remote_io"` option in a context).

Comment (madsbk) [reply python/cudf/cudf/io/parquet.py:625]: Good point, removed

File: python/cudf/cudf/options.py [hunk 4]
```diff
@@ -351,6 +351,20 @@ def _integer_and_none_validator(val):
...
+_register_option(
+    "kvikio_remote_io",
+    _env_get_bool("CUDF_KVIKIO_REMOTE_IO", False),
+    textwrap.dedent(
+        """
+        Whether to use KvikIO's remote IO backend or not.
+        \tWARN: this is experimental and may be removed at any time
+        \twithout warning or deprecation period.
+        \tValid values are True or False. Default is False.
+    """
```
Comment (rjzamora) [inline python/cudf/cudf/options.py:363]: Is this a good place to add a note about `KVIKIO_NTHREADS`? (That var seems critical to performance)

Comment (madsbk) [reply python/cudf/cudf/options.py:365]: added a note

File: cpp/src/io/utilities/datasource.cpp [hunk 5]
```diff
@@ -390,6 +392,84 @@ class user_datasource_wrapper : public datasource {
...
+    this->host_read(offset, count, h_data.data());
+    return datasource::buffer::create(std::move(h_data));
+  }
+
+  /**
+   * @brief Is `url` referring to a remote file supported by KvikIO?
+   */
+  static bool is_supported_remote_url(std::string const& url)
+  {
+    // Regular expression to match "<s3|http|https>://"
```
Comment (pmattione-nvidia) [inline cpp/src/io/utilities/datasource.cpp:464]: Wondering about long-term: If we're given an http or https URI, how would we tell whether that's on amazon s3 vs some other cloud service?

Comment (pmattione-nvidia) [reply cpp/src/io/utilities/datasource.cpp:464]: E.g. if users rely on https automatically working for s3, then we need to change how https works in the future, could this break.

File: cpp/src/io/utilities/datasource.cpp [hunk 6]
```diff
@@ -390,6 +392,84 @@ class user_datasource_wrapper : public datasource {
...
+  }
+
+ public:
+  explicit remote_file_source(char const* filepath) : _kvikio_file{create_s3_endpoint(filepath)} {}
+
+  ~remote_file_source() override = default;
+
+  [[nodiscard]] bool supports_device_read() const override { return true; }
+
+  [[nodiscard]] bool is_device_read_preferred(size_t size) const override { return true; }
```
Comment (vuule) [inline cpp/src/io/utilities/datasource.cpp:412]: Should this be `true`? the kvikIO reads to host and then copies to device. Is there any benefit over letting the caller manage how the H2D copy is done?

**Review by vuule — COMMENTED** [review]
great stuff!
few small comments

File: cpp/src/io/utilities/datasource.cpp (hunk 6 above)
Comment (madsbk) [reply cpp/src/io/utilities/datasource.cpp:412]: Yes, this enable KvikIO to optimize the communication like in  https://github.com/rapidsai/kvikio/pull/519

File: cpp/src/io/utilities/datasource.cpp (hunk 5 above)
Comment (madsbk) [reply cpp/src/io/utilities/datasource.cpp]: Good point, I have limited the support to `s3://` urls for now. 
KvikIO also supports [regular http servers](https://github.com/rapidsai/kvikio/blob/branch-24.12/cpp/include/kvikio/remote_handle.hpp#L145) but let's address that in a later PR.

**Review by pmattione-nvidia — APPROVED** [review]

**Review by vuule — APPROVED** [review]

File: cpp/src/io/utilities/datasource.cpp [hunk 7]
```diff
@@ -389,6 +391,86 @@ class user_datasource_wrapper : public datasource {
...
+
+  std::future<size_t> device_read_async(size_t offset,
+                                        size_t size,
+                                        uint8_t* dst,
+                                        rmm::cuda_stream_view stream) override
+  {
+    CUDF_EXPECTS(supports_device_read(), "Device reads are not supported for this file.");
+
+    auto const read_size = std::min(size, this->size() - offset);
+    return _kvikio_file.pread(dst, read_size, offset);
```
Comment (shrshi) [inline cpp/src/io/utilities/datasource.cpp:423]: Question: If insufficient memory is allocated in `dst`, does `pread` catch the exception thrown?

**Review by shrshi — APPROVED** [review]
Overall looks good to me!

File: cpp/src/io/utilities/datasource.cpp (hunk 7 above)
Comment (madsbk) [reply cpp/src/io/utilities/datasource.cpp:423]: KvikIO has no overflow protection, it is undefined behavior if `dst` is too small.

**Comment by madsbk** [comment]
@rjzamora do you have anything else?

**Review by rjzamora — APPROVED** [review]
Thanks for this @madsbk (very exciting)!

**Review by galipremsagar — APPROVED** [review]

**Comment by madsbk** [comment]
Thanks all!

**Pull request merged**